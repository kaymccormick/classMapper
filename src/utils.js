"use strict";
exports.__esModule = true;
// eslint-disable-next-line @typescript-eslint/no-unused-vars
var ast_types_1 = require("ast-types");
var immutable_1 = require("immutable");
function copyTree(node) {
    var out = immutable_1.Map();
    ast_types_1.eachField(node, function (name, value) {
        if (Array.isArray(value)) {
            if (typeof value[0] === 'string') { //instanceof namedTypes.Node) {
                throw new Error('');
            }
        }
        else if (value
            && value.constructor && value.constructor.name === "Node") {
            out = out.set(name, copyTree(value));
        }
        else if (value && value.type) {
            throw new Error(value.type);
        }
        else {
            {
                out = out.set(name, value);
            }
        }
    });
    return out;
}
exports.copyTree = copyTree;
