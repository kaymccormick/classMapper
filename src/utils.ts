// eslint-disable-next-line @typescript-eslint/no-unused-vars
import {getFieldNames, getFieldValue, eachField, namedTypes, builders} from 'ast-types';
import { CommentKind } from 'ast-types/gen/kinds';
import { Map,List } from 'immutable';
import { ValueKind, BasicLogger } from './types';
export { copyTree, CopyTreeResult } from './copyTree.prune';

import { NodePath } from 'ast-types/lib/node-path';

export const promises: Promise<any>[] = [];
export function p<T>(promise: Promise<T>): Promise<T> {
    promises.push(promise);
    return promise;
}



export function normalizeDocComment(report: any, lines: string[]): string[] {
    let newLines = lines.map(line => line.replace(/^\s*\*\s*/, ''));
    if(newLines[0].trim()) {
        newLines.splice(0, 0, '');
    }
    if(newLines[newLines.length - 1].trim()) {
        newLines.splice(newLines.length, 0, ' ');
    } else {
        newLines[newLines.length - 1] = ' ';
    }

    newLines = newLines.map((line, index) => index === 0 ? `*` : index === newLines.length -1 ? line : ` * ${line}`);
    return newLines;
}

/**
 * Purpose: ?
 */
export function processComments(report: any, nodePath: NodePath<namedTypes.Node>) {
    const node = nodePath.node;
    let column: number|undefined;
    if(node.loc) {
        column = node.loc.start.column;
    }
    report(`${node.type} ${column}`);

    const comments = node.comments;
    if(!comments || !comments.length) {
        return;
    }
    comments.forEach((comment: CommentKind): void => {
        const val = comment.value;
        report(val.replace(/\n/g, '\\n'));
        let containsAtUuid = val.indexOf('@uuid') !== -1;
        const lines = val.split(/\r?\n/g);
        let docComment = false;
        let singleLine = false;
        if(lines[0] === '*') {
            docComment = true;
        }
        if(lines.length === 1) {
            singleLine = true;
        }
        //report(`${singleLine} ${docComment} ${containsAtUuid}`);
        if(containsAtUuid || docComment) {
            if(!containsAtUuid) {
                report('here');
            }
            const newLines = normalizeDocComment(report, lines);
            const newVal = newLines.join('\n');
            //  report(`processComments: ${newVal}\n.\n`);
            comment.value = newVal;
        }
    });
}
