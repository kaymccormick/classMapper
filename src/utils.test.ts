import { builders as b, namedTypes as n } from 'ast-types';
import { copyTree } from './utils';

function makeFile() {
    return b.file(b.program([]));
}

test('1', (): void => {
    const file =  makeFile();
    expect(n.File.check(file)).toBeTruthy();
    const tree = copyTree(file);
    expect(tree.toJS()).toMatchSnapshot();
});
