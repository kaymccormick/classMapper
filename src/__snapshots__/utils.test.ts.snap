// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 1`] = `
Object {
  "name": null,
  "program": Object {
    "body": Array [],
    "directives": Array [],
    "interpreter": null,
    "type": "Program",
  },
  "type": "File",
}
`;
